src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c

```diff
@@ -0,0 +1,223 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Xilinx CU driver for memory to memory BO copy
+ *
+ * Copyright (C) 2020 Xilinx, Inc.
+ *
+ * Authors: David Zhang <davidzha@xilinx.com>
+ */
+
+#include <linux/types.h>
+#include <linux/uaccess.h>
+#include <linux/vmalloc.h>
+#include <linux/slab.h>
+#include <linux/cdev.h>
+#include <linux/fs.h>
+#include <linux/device.h>
+#include <linux/io.h>
+#include <linux/ioctl.h>
+
+#include "../xocl_drv.h"
+#include "../userpf/common.h"
+#include "xrt_cu.h"
+
+/* IOCTL interfaces */
+#define XOCL_M2M_MAGIC 0x4d324d // "M2M"
+
+#define	M2M_ERR(m2m, fmt, arg...)	\
+	xocl_err(&(m2m)->m2m_pdev->dev, fmt "\n", ##arg)
+#define	M2M_WARN(m2m, fmt, arg...)	\
+	xocl_warn(&(m2m)->m2m_pdev->dev, fmt "\n", ##arg)
+#define	M2M_INFO(m2m, fmt, arg...)	\
+	xocl_info(&(m2m)->m2m_pdev->dev, fmt "\n", ##arg)
+#define	M2M_DBG(m2m, fmt, arg...)	\
+	xocl_dbg(&(m2m)->m2m_pdev->dev, fmt "\n", ##arg)
+
+/* This is the real register map for the copy bo CU */
+struct start_copybo_cu_cmd {
+  uint32_t src_addr_lo;      /* low 32 bit of src addr */
+  uint32_t src_addr_hi;      /* high 32 bit of src addr */
+  uint32_t src_bo_hdl;       /* src bo handle, cleared by driver */
+  uint32_t dst_addr_lo;      /* low 32 bit of dst addr */
+  uint32_t dst_addr_hi;      /* high 32 bit of dst addr */
+  uint32_t dst_bo_hdl;       /* dst bo handle, cleared by driver */
+  uint32_t size;             /* size in bytes */
+};
+
+struct xocl_m2m {
+	struct platform_device 	*m2m_pdev;
+	struct xrt_cu 		m2m_cu;
+	struct mutex 		m2m_lock;
+};
+
+static int copy_bo(struct platform_device *pdev, uint64_t src_paddr,
+	uint64_t dst_paddr, uint32_t src_bo_hdl, uint32_t dst_bo_hdl,
+	uint32_t size)
+{
+	struct xocl_m2m *m2m = platform_get_drvdata(pdev);
+	xdev_handle_t xdev = xocl_get_xdev(m2m->m2m_pdev);
+	struct xrt_cu *xcu = &m2m->m2m_cu;
+	struct start_copybo_cu_cmd cmd;
+
+	cmd.src_addr_lo = (uint32_t)src_paddr;
+	cmd.src_addr_hi = (src_paddr >> 32) & 0xFFFFFFFF;
+	cmd.dst_addr_lo = (uint32_t)dst_paddr;
+	cmd.dst_addr_hi = (dst_paddr >> 32) & 0xFFFFFFFF;
+	cmd.src_bo_hdl = src_bo_hdl;
+	cmd.dst_bo_hdl = dst_bo_hdl;
+	cmd.size = size;
```

Comment:
cmd.size = size / KDMA_DATA_WIDTH;

Reply:
Oops, the argument size is already calculated with bus width in xocl_bo.c. Please add a comment to let people know cmd.size is not in bytes.

Reply:
Yes, fixed it with additional comments